iree/compiler/Dialect/HAL/Target/LLVM/internal/RiscvLinkerTool.cpp

```diff
@@ -71,13 +71,7 @@ class RiscvLinkerTool : public LinkerTool {
 
     // Statically link all dependencies so we don't have any runtime deps.
     // We cannot have any imports in the module we produce.
-    // flags.push_back("-static");
-
-    // HACK: we insert mallocs and libm calls. This is *not good*.
-    // We need hermetic binaries that pull in no imports; the MLIR LLVM
-    // lowering paths introduce a bunch, though, so this is what we are
-    // stuck with.
-    flags.push_back("-shared");
+    flags.push_back("-static");
```

Comment:
Seeing CI failures when running this test: https://github.com/google/iree/blob/main/build_tools/kokoro/gcp_ubuntu/cmake/linux/riscv64/test.sh

logs: https://source.cloud.google.com/results/invocations/18dc0414-3cb1-423e-8c08-e706fe7b344e/targets/iree%2Fgcp_ubuntu%2Fcmake%2Flinux%2Friscv64%2Fpresubmit/log

```
Test e2e mlir --> bytecode module --> iree-run-module
[19:09:47 UTC] echo 'Test e2e mlir --> bytecode module --> iree-run-module'
[19:09:47 UTC] /tmpfs/src/github/iree/build-host/install/bin/iree-translate -iree-input-type=mhlo -iree-mlir-to-vm-bytecode-module -iree-hal-target-backends=dylib-llvm-aot -iree-llvm-link-embedded=false -iree-llvm-target-triple=riscv64 -iree-llvm-target-cpu=generic-rv64 -iree-llvm-target-cpu-features=+m,+a,+f,+d,+c -iree-llvm-target-abi=lp64d /tmpfs/src/github/iree/iree/tools/test/iree-run-module.mlir -o /tmpfs/src/github/iree/build-riscv/iree-run-module-llvm_aot.vmfb
/usr/src/toolchain_iree/bin/riscv64-unknown-linux-gnu-ld: warning: cannot find entry symbol _start; defaulting to 0000000000010120
[[19:09:47 UTC] /usr/src/qemu-riscv/qemu-riscv64 -cpu rv64,x-v=true,x-k=true,vlen=256,elen=64,vext_spec=v1.0 -L /usr/src/toolchain_iree/sysroot /tmpfs/src/github/iree/build-riscv/iree/tools/iree-run-module --driver=dylib --module_file=/tmpfs/src/github/iree/build-riscv/iree-run-module-llvm_aot.vmfb --entry_function=abs --function_input=f32=-10
I ../iree/tools/utils/vm_util.cc:185] Creating driver and device for 'dylib'...
Error in cpuinfo: processor architecture is not supported in cpuinfo
../iree/base/internal/dynamic_library_posix.c:183: NOT_FOUND; dynamic library not found on any search path; while invoking native function hal.executable.create; while calling import;
[ 1]   native hal.executable.create:0 -
[ 0] bytecode module.__init:230 /tmpfs/src/github/iree/iree/tools/test/iree-run-module.mlir:7:13
      at /tmpfs/src/github/iree/iree/tools/test/iree-run-module.mlir:6:1; creating context
[19:09:48 UTC] IREE_RUN_OUT=
```

@hcindyl any suggestions for changing the code here or changing what the test is doing?

Reply:
we could probably eject this riscv-specific linker tool now that we have the embedded stuff on by default - feels likely to bitrot and not sure there are many uses for it - open to hearing if anyone is using it for anything

Reply:
hmm... does the PR work on X86_64 with ld as the linker? There's nothing special on the test script, only the use of riscv32 clang as the linker.

I am okay with dropping the riscv-specific linker tool all together, but want to make sure the `iree-run-module` test here still works.

Reply:
The Embedded, Unix, Windows, Android, and Wasm linker tools work, so I'm trying to figure out if I broke the Riscv linker.

The new regression test fails with the embedded and Windows linkers if the musl bitcode is not inserted. The Android linker suspiciously succeeds... maybe that fails at runtime though?

Reply:
I think musl is used to resolve the issue of unresolved symbols. In Android it would be deferred to dl_resolve, so the error may not show up until runtime.

Reply:
Yeah, then it might make sense to turn the new tests into `check` tests so they execute on Android.

Reply:
>_> the https://github.com/google/iree/blob/main/iree/test/e2e/regression/BUILD file is such a mess already though

Reply:
If we replace https://github.com/google/iree/blob/main/build_tools/kokoro/gcp_ubuntu/cmake/linux/riscv64/test.sh#L41 with `-iree-llvm-embedded-linker-path=${BUILD_HOST_DIR?}/install/bin/lld` to use the EmbeddedLinkerTool path, I think we can remove RiscvLinkerTool. Sorry I should have noticed it earlier.

Reply:
Nice, that works for me. I was wondering if we had a specific reason to keep the RiscvLinkerTool around. It would be a good spot to add platform specific information or file formats that could help with debugging, but I guess we're not planning that?

Reply:
Thankfully for any elf-compatible linux-y systems the files we produce with the embedded linker tool can be loaded with dlopen - so we can have our embedded linker tool spit out dwarf symbols and then use the system loader to dlopen them. So I think the only reason to have a specific linker tool would be if we were targeting systems that didn't have that (windows/mac) - and I've had some ideas about plumbing through a DWARF->PDB converter so we could load the elfs with our loader and then resolve symbols with the platform tools... muahaha